openforcefield/topology/molecule.py

```diff
@@ -3176,6 +3364,319 @@ def from_openeye(oemol, allow_undefined_stereo=False):
         return toolkit.from_openeye(
             oemol, allow_undefined_stereo=allow_undefined_stereo)
 
+    def to_qcschema(self, multiplicity=1, conformer=0):
+        """
+        Generate the qschema input format used to submit jobs to archive
+        or run qcengine calculations locally
+        spec can be found here <https://molssi-qc-schema.readthedocs.io/en/latest/index.html>
+
+        Parameters
+        ----------
+        multiplicity : int, default=1, the multiplicity of the molecule required for qcschema
+
+        conformer : int, default=0, the index of the conformer that should be used for qcschema
+
+        Returns
+        ---------
+        qcelemental.models.Molecule : a validated qcschema
+
+        Example
+        -------
+
+        Create and validate a qcelemental input
+
+        >>> import qcelemental as qcel
+        >>> mol = Molecule.from_smiles('CC')
+        >>> mol.generate_conformers(n_conformers=1)
+        >>> qcschema = mol.to_qcschema()
+
+        Raises
+        --------
+        ImportError : if qcelemental is not installed; the qcschema can not be validated.
+
+        InvalidConformerError : if there is no conformer found at the given index.
+        """
+
+        try:
+            import qcelemental as qcel
+        except ImportError:
+            raise ImportError('Please install QCElemental via conda install -c conda-forge qcelemental '
+                              'to validate the schema')
+
+        try:
+            geometry = self.conformers[conformer].in_units_of(unit.bohr)
+        except (IndexError, TypeError):
+            raise InvalidConformerError('The molecule must have a conformation to produce a valid qcschema; '
+                                        f'no conformer was found at index {conformer}.')
+
+        # Gather the required qschema data
+        charge = sum([atom.formal_charge for atom in self.atoms])
+        connectivity = [(bond.atom1_index, bond.atom2_index, bond.bond_order) for bond in self.bonds]
+        symbols = [Element.getByAtomicNumber(atom.atomic_number).symbol for atom in self.atoms]
+
+        schema_dict = {'symbols': symbols, 'geometry': geometry, 'connectivity': connectivity,
+                       'molecular_charge': charge, 'molecular_multiplicity': multiplicity}
+
+        return qcel.models.Molecule.from_data(schema_dict, validate=True)
+
+    @classmethod
+    def from_mapped_smiles(cls, mapped_smiles, allow_undefined_stereo=False):
```

Comment:
Does this function (intend to) handle cases where partial maps exist, e.g. torsiondrives where only the 4 driven atoms have an index?

Reply:
No only fully mapped smiles would work here due to the reordering requiring a new index for each atom, I could change the name to further state this? `from_fully_mapped_smiles` ?

Reply:
I don't see why partial ordering wouldn't work, but at the same time as a user I could take my partial ordering, and then just do a sequential map on the other atoms. Not a big deal, but something this function could also probably do.

For now I would keep the original name as it is more intuitive to search `from_map` in the API, and then just throw an exception if anything is mapped to 0, indicating a partial map.

Reply:
Yeah, I think we could support that in future and probably even offer a way of generating the partial maps using something like the isomorphic checking function.

Reply:
I'm mostly with Josh here -- Arbitrarily ordering non-indexed atoms would be a dangerous behavior, where we could quickly implement it in some arbitrary way, but then some user would make workflows reliant on that arbitrary algorithm, and we'd either get angry emails when we change the ordering later, or we'd have to maintain this arbitrary behavior forever.